### PR TITLE
Benchmark branch

### LINKED ISSUE #181: HTM.java Needs Baseline Benchmarks
HTM.java could use extensive Benchmark apparatus utilizing jmh to form a baseline upon which optimization can then be done.

### PR DESCRIPTION
**Pull request by cogmission**
Fixes #181 

Added optimizations to classes.
Added *NEW* benchmark package and implemented new Benchmarks for the SpatialPooler and TemporalMemory.
Added some documentation fixes.

### CODE REVIEW

**Comment by cogmission** [comment]
@rhyolight Just a gentle reminder. :)

File: src/jmh/java/org/numenta/nupic/benchmarks/AbstractAlgorithmBenchmark.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+	
+	/**
+	 * Create and return a {@link Parameters} object.
+	 * 
+	 * @return
+	 */
+	protected Parameters getParameters() {
+    	Parameters parameters = Parameters.getAllDefaultParameters();
+    	parameters.setParameterByKey(KEY.INPUT_DIMENSIONS, new int[] { 8 });
+        parameters.setParameterByKey(KEY.COLUMN_DIMENSIONS, new int[] { 2048 });
```
Comment (rhyolight) [inline src/jmh/java/org/numenta/nupic/benchmarks/AbstractAlgorithmBenchmark.java]: Indentation problem on this and subsequent lines.

File: src/jmh/java/org/numenta/nupic/benchmarks/SpatialPoolerLocalInhibitionBenchmark.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+	public void init() {
+		super.init();
+		
+		input = new int[7][8];
+    	for(int i = 0;i < 7;i++) {
+    		input[i] = encoder.encode((double) i + 1);
+    	}
+	}
+	
+	@Benchmark
```
Comment (rhyolight) [inline src/jmh/java/org/numenta/nupic/benchmarks/SpatialPoolerLocalInhibitionBenchmark.java]: Indentation.

File: src/jmh/java/org/numenta/nupic/benchmarks/TemporalMemoryBenchmark.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+	    	for(int i = 0;i < 7;i++) {
+	    		temporalMemory.compute(memory, SDRs[i], true);
+	    	}
+	    }
+	}
+	
+	@Benchmark
+	@BenchmarkMode(Mode.AverageTime)
+    @OutputTimeUnit(TimeUnit.MILLISECONDS)
+    public ComputeCycle measureAvgCompute_7_Times(Blackhole bh) throws InterruptedException {
```
Comment (rhyolight) [inline src/jmh/java/org/numenta/nupic/benchmarks/TemporalMemoryBenchmark.java]: Indentation.

**Comment by rhyolight** [comment]
Aside from a few indentation discrepancies, LGTM. :+1:

**Comment by cogmission** [comment]
Thanks Matt. I'll fix the indentation

**Pull request merged**